### PR TITLE
Add the option to put the file name in the returned path

### PR DESCRIPTION
**Pull request by olivier-deschenes**
Added the option to put the file name in the path

Two new configurations:
 - `putFileNameInPath` (*Boolean*) : True if the file name should be added, else false. Defaults to false.
 - `prefixSeparator` (*String*) : The separator to put between the prefix and the path. Defaults to `:`.

### CODE REVIEW

**Comment by nidu** [comment]
Hi @olivier-deschenes , thanks for the PR.
Could you please tell what's the use case for that?
I just wonder if controlling adding a file name should be a preference.

I would think that you might have different tools in your project. Some of them may need file name as part of the key and some of them may not. Switching between 2 approaches in preferences would've been an overkill. What do you think of instead of having a `putFileNameInPath` the extension would expose two commands - "Copy Json Path" and "Copy Json Path with File Name"?

**Comment by olivier-deschenes** [comment]
Hi @nidu,

Thanks for your plugin btw, works great.

I use your plugin for [react-i18next](https://github.com/i18next/react-i18next) so I can copy the translation path easily in my react code (the namespace of my translation key is the json file name).

I think that another command for this use case is maybe a little overkill, because I do not see a use case where a user would need the file name only _sometimes_. He needs it (like me 😁) or not. But I would not mind at all another command. 

I'll let you decide 🫡

File: src/extension.ts [hunk 1]
```diff
@@ -37,7 +47,15 @@ export function activate(context: vscode.ExtensionContext) {
...
             const text = editor.document.getText()
             // JSON.parse(text)
-            const path = jsPathTo(text, editor.document.offsetAt(editor.selection.active), nonQuotedKeyRegex)
+            let path = jsPathTo(text, editor.document.offsetAt(editor.selection.active), nonQuotedKeyRegex)
+
+            if (fileNameAsPrefix && editor.document.fileName) {
+                const fileName = parse(editor.document.fileName).base;
+                const fileNameWithoutExtension = fileName.substring(0, fileName.lastIndexOf('.'))
+                
+                path = `${fileNameWithoutExtension}${prefixSeparator}${path}`
```
Comment (nidu) [inline src/extension.ts:56]: When calling the function in a not yet saved file - you still get a prefix separator, e.g. `:attributes[0].value`. Is it desirable?

File: src/extension.ts [hunk 2]
```diff
@@ -37,7 +47,15 @@ export function activate(context: vscode.ExtensionContext) {
 
             const text = editor.document.getText()
             // JSON.parse(text)
-            const path = jsPathTo(text, editor.document.offsetAt(editor.selection.active), nonQuotedKeyRegex)
+            let path = jsPathTo(text, editor.document.offsetAt(editor.selection.active), nonQuotedKeyRegex)
+
+            if (fileNameAsPrefix && editor.document.fileName) {
+                const fileName = parse(editor.document.fileName).base;
+                const fileNameWithoutExtension = fileName.substring(0, fileName.lastIndexOf('.'))
```
Comment (nidu) [inline src/extension.ts:54]: What if the file doesn't have an extension? For example prettier can be configured with JSON `.prettierrc` file (https://prettier.io/docs/en/configuration.html). In this case file name is empty. I expect there can be similar JSON files not starting with a dot as well. Could you take these cases into account?

**Review by nidu — CHANGES_REQUESTED** [review]

File: package.json [hunk 3]
```diff
@@ -45,6 +45,16 @@
           "type": "string",
           "default": "^[a-zA-Z$_][a-zA-Z\\d$_]*$",
           "markdownDescription": "Regex that tests whether key in path can be used without quotes. If key matches - key is not quoted. Use it if you want to have path `a.b-c` instead of `a[\"b-c\"]` for example. If you want all keys to be escaped - use regex that doesn't match anything, e.g. single space ` ` will only match single space, which is very rare object key (you can go more complicated like `a{1000}`). Default is `^[a-zA-Z$_][a-zA-Z\\d$_]*$`."
+        },
+        "extension.copyJsonPath.putFileNameInPath": {
+          "type": "boolean",
+          "default": "false",
```
[the remark this answers is not in the document]
Comment (olivier-deschenes) [reply package.json]: Good catch, my mistake.

File: src/extension.ts (hunk 1 above)
Comment (olivier-deschenes) [reply src/extension.ts:56]: Good point, I added a check for this case.

File: src/extension.ts (hunk 2 above)
Comment (olivier-deschenes) [reply src/extension.ts]: Fixed

**Comment by olivier-deschenes** [comment]
@nidu Thanks for the feedback, the issues should be resolved.

**Comment by nidu** [comment]
@olivier-deschenes thanks for the changes! Just the last one - removing dot from the file name starting with dot seems a bit strange to me. When we have extension - we remove extension which makes sense. But when we have dot - I'd consider the dot a part of the file name. I think this use case might not even ever be used, but this little quirk sounds like unnecessary mental addition. What do you think?

**Comment by olivier-deschenes** [comment]
@nidu , I do agree that this use case will probably never be used 😁, but if we leave the `.`, the path will start with a `.` thus making the entire path not very valid for anything.

(any sort of parser will probably not handle the dot correcly at the start)

**Comment by nidu** [comment]
@olivier-deschenes yeap, this is a weird use case, but without the dot we get a non existing file name and it's not missing an extension (which I could guess probably and if I see a file name without an extension I'd consider it's JSON probably), but it's missing the first character of file name, which I would never think of myself 😄

**Comment by olivier-deschenes** [comment]
Fixed. I use directly the `name` provided by `parse` 😊

**Comment by nidu** [comment]
Perfect, thanks!

**Comment by nidu** [comment]
Published version 0.3.0

**Pull request merged**